internal/hostport/hostport_nftables.go

```diff
@@ -0,0 +1,308 @@
+			"inet_proto", ".", "inet_service", ":", ipaddr, ".", "inet_service",
+		),
+		Comment: knftables.PtrTo("hostports on all local IPs (protocol . hostPort -> podIP . podPort)"),
+	})
+	tx.Add(&knftables.Map{
+		Name: hostIPPortsMap,
+		Type: knftables.Concat(
+			ipaddr, ".", "inet_proto", ".", "inet_service", ":", ipaddr, ".", "inet_service",
+		),
+		Comment: knftables.PtrTo("hostports on specific IPs (hostIP . protocol . hostPort -> podIP . podPort)"),
+	})
+
+	// Create the "hostports" chain with the map lookup rules, and then create
+	// "prerouting" and "output" chains that call the "hostports" chain for
+	// locally-destined packets.
+	tx.Add(&knftables.Chain{
+		Name: "hostports",
+	})
+	tx.Flush(&knftables.Chain{
+		Name: "hostports",
+	})
+	// hostIPPortsMap check must come first since the hostPortsMap rule catches all IPs
+	tx.Add(&knftables.Rule{
+		Chain: "hostports",
+		Rule: knftables.Concat(
+			"dnat", ip, "addr . port to",
+			ip, "daddr", ".", "meta l4proto", ".", "th dport", "map", "@", hostIPPortsMap,
+		),
+	})
+	tx.Add(&knftables.Rule{
+		Chain: "hostports",
+		Rule: knftables.Concat(
+			"dnat", ip, "addr . port to",
+			"meta l4proto", ".", "th dport", "map", "@", hostPortsMap,
+		),
+	})
+
+	tx.Add(&knftables.Chain{
+		Name:     "prerouting",
+		Type:     knftables.PtrTo(knftables.NATType),
+		Hook:     knftables.PtrTo(knftables.PreroutingHook),
+		Priority: knftables.PtrTo(knftables.DNATPriority),
+	})
+	tx.Flush(&knftables.Chain{
+		Name: "prerouting",
+	})
+	tx.Add(&knftables.Rule{
+		Chain: "prerouting",
+		Rule:  "fib daddr type local  goto hostports",
+	})
+
+	tx.Add(&knftables.Chain{
+		Name:     "output",
+		Type:     knftables.PtrTo(knftables.NATType),
+		Hook:     knftables.PtrTo(knftables.OutputHook),
+		Priority: knftables.PtrTo(knftables.DNATPriority),
+	})
+	tx.Flush(&knftables.Chain{
+		Name: "output",
+	})
+	tx.Add(&knftables.Rule{
+		Chain: "output",
+		Rule:  "fib daddr type local  goto hostports",
+	})
+
+	// Create the "masquerading" chain, linked to the postrouting hook.
+	tx.Add(&knftables.Chain{
+		Name:     "masquerading",
+		Type:     knftables.PtrTo(knftables.NATType),
+		Hook:     knftables.PtrTo(knftables.PostroutingHook),
+		Priority: knftables.PtrTo(knftables.SNATPriority),
+	})
+	tx.Flush(&knftables.Chain{
+		Name: "masquerading",
+	})
+	// You can't write an nftables rule that checks "source IP == destination IP" so
+	// instead we match it by looking up the source and destination IP in a set that
+	// has been filled in (by HostportManager) with entries containing the same pod IP
+	// twice. We can't _exactly_ match only the traffic that was definitely DNATted by
+	// our rule as opposed to someone else's, but if the traffic has been DNATted and
+	// has src=dst=podIP then _someone_ needs to masquerade it.
+	tx.Add(&knftables.Set{
+		Name: hairpinSet,
+		Type: knftables.Concat(
+			ipaddr, ".", ipaddr,
+		),
+		Comment: knftables.PtrTo("hostport hairpin connections"),
+	})
```

Comment:
I never remember, but IIRC sets has different semantics and is easy to wipe them out when just trying to add a single element, can't say this is correct or wrong, just you should double check adding  a new  hostport does not delete all the previous set entries

Reply:
"Add" never erases anything. It does one of:
1. create the object if it doesn't exist.
2. no-op if the object already exists and matches what you tried to create.
3. error if the object already exists and is different from what you tried to create.

As noted in a comment above, for the chains I do Add+Flush, but for the sets just Add.